_posts/2023-11-01-automated-test-generation-chess-puzzles-with-dafny.markdown

```diff
@@ -0,0 +1,279 @@
+    var board := Board([Piece(kind:=King(White)  , at:=Pos(row:=6,col:=2)),
+                        Piece(kind:=Knight(Black), at:=Pos(row:=6,col:=7)),
+                        Piece(kind:=Knight(Black), at:=Pos(row:=5,col:=3)),
+                        Piece(kind:=Pawn(Black)  , at:=Pos(row:=3,col:=6)),
+                        Piece(kind:=Pawn(Black)  , at:=Pos(row:=5,col:=5))]);
+    Describe(board);
+  }
+  method {:test} Test1() {
+    var board := Board([Piece(kind:=King(White)  , at:=Pos(row:=0,col:=7)),
+                        Piece(kind:=Knight(Black), at:=Pos(row:=1,col:=5)),
+                        Piece(kind:=Knight(Black), at:=Pos(row:=1,col:=4)),
+                        Piece(kind:=Pawn(Black)  , at:=Pos(row:=0,col:=5)),
+                        Piece(kind:=Pawn(Black)  , at:=Pos(row:=0,col:=6))]);
+    Describe(board);
+  }
+}
+{% endhighlight %}
+</div><!--.file-->
+
+Note also that Dafny annotates the two test methods with the `{:test}` attribute. This means that after saving the tests to a file (`tests.dfy`), you can then use the `dafny test tests.dfy` command to compile and execute them. The default compilation target is C# but you can pick a different one using the `--target` option - read more about the `dafny-test` command [here](https://dafny.org/dafny/DafnyRef/DafnyRef#sec-dafny-test).
+
+Oftentimes, you would want to translate the tests from Dafny code to some other format that you are using throughout your project. To do that, you can serialize every test by adding a set of print statements at the end of the test entry method. For our purposes, I wrote the `SerializeToSVG` method (included in the <a href="/blog/assets/src/chess.dfy">sources</a>) that converts each test to an `.svg` image of the corresponding chess board. You can download the relevant image assets <a href="/blog/assets/images/test-generation/assets">here</a> (originally from <a href="https://github.com/tabler/tabler-icons/tree/master">tabler icons</a>). Let us look at the results in picture form (knights are the pieces that look like a horse’s head and the white king is the only piece that isn’t filled in):
+
+<center>
+{% include test-generation-svgs/block-no-inlining/0.svg %}
+{% include test-generation-svgs/block-no-inlining/1.svg %}
+</center>
+
+The image on the right shows a checkmate and the image on the left corresponds to the absence of either a checkmate or a check. While the exact configurations will depend on the version of Dafny you are using, you will always get two tests and there will never be a test for the case in which the king is checked but can escape. This is because under the block coverage criteria, such a test is unnecessary - there is no one statement in the target method that this test would cover that the two existing ones do not. To expand the set of tests Dafny generates we can instead prompt it to target path-coverage - the most expensive form of coverage Dafny supports. The relevant command is
```

Comment:
Have paraphrased the paragraph to say that Dafny could have generated three (or even more) tests but that it is not compelled to do so by the selected coverage criterion. Two tests are sufficient because they cover all statements within the test entry method (and because we are not inlining any other methods, so Dafny only cares about statements within the test entry method). It will always generate two tests in the particular case due to how test generation is optimized under the hood but it probably doesn't make sense to go into that in the blogpost.